[[ Bug 21993 ]] Tweak __MCName and __MCString layouts

runrevmark opened the pull request:
This patch tweaks both the __MCName and __MCString struct layouts
to ensure they are more efficient.

The __MCName struct has had the now unused 'hash' member removed
when compiling for 64-bit architectures as this is stored in
free bits elsewhere in the structure.

The __MCString struct has had its layout changed when compiling
for 32-bit architectures so that the numeric_value double is
8-byte aligned, a padding member to ensure the struct is 32-bytes
in size on all platforms has also been added.

Additionally a low-level memory test has been added which verifies
the sizes of all the core __MCValue derived structs are the
expected (hand calculated and optimized for) values.

livecodeali commented:
@livecode-vulcan review ok 019b0fa

livecode-vulcan commented:
💙 review by @livecodeali ok 019b0fa8211405739c70a38d95972bb63a7478a5

livecode-vulcan commented:
😞 test failure 019b0fa8211405739c70a38d95972bb63a7478a5

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/409)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/401)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/401)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/401)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/401)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1923)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/151)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/150)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1954)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1944)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1943)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1972)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1913)

livecodeali commented:
@livecode-vulcan retry 019b0fa

runrevmark commented:
@livecodeali This actually failed on win32 because a test failed - on 32-bit windows the __MCString struct is 32 bytes and not 28 it seems. It might be that the default compiler settings these days on 32-bit win are to make struct sizes 8-byte aligned. In any case, I'll fix - the main thing is that the value structs are as close to a power-of-2 as possible as they have the best chance of being allocated with minimal wastage (not that the Windows 32-bit allocator does that, but hey - that's a different problem!).

runrevmark commented:
@livecodeali : I've updated this now - the reason for the difference on win32 is because the double in __MCString was not 8-byte aligned. I've tweaked the layout of the struct to make it so now (moving capacity after it).

livecodeali commented:
@livecode-vulcan review ok cda348f

livecode-vulcan commented:
💙 review by @livecodeali ok cda348f2fd4d34af605b0f30547da5ea4cb85cd7

runrevmark commented:
@livecode-vulcan review not ok cda348f

livecode-vulcan commented:
💔 review by @runrevmark not ok cda348f2fd4d34af605b0f30547da5ea4cb85cd7

runrevmark commented:
@livecodeali Updated again - the layouts of __MCString need to be different from 32/64 bit - I remembered to run the foundation tests locally on mac in 32/64 bit mode this time... So it should pass this time on Travis :)

livecodeali commented:
@livecode-vulcan review ok b13c8a1

livecode-vulcan commented:
💙 review by @livecodeali ok b13c8a1aef5c4e898e67062cb3332091f7b40399

livecode-vulcan commented:
😞 test failure b13c8a1aef5c4e898e67062cb3332091f7b40399

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/442)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/434)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/434)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/434)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/434)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1956)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/184)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/183)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1987)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1977)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1976)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2001)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1942)

livecodeali commented:
@livecode-vulcan retry b13c8a1

livecode-vulcan commented:
😞 test failure b13c8a1aef5c4e898e67062cb3332091f7b40399

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/445)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/437)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/437)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/437)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/437)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1959)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/187)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/186)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1990)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1980)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1979)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2007)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1948)

livecodeali commented:
@livecode-vulcan retry b13c8a1

livecode-vulcan commented:
😞 test failure b13c8a1aef5c4e898e67062cb3332091f7b40399

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/450)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/442)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/442)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/442)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/442)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1964)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/192)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/191)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1995)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1985)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1984)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2012)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1953)

runrevmark commented:
Hah! Okay so the size of that struct is going to be 32-bytes :) The most recent adjustment made it so that the 4-byte padding was moved from before the 'numeric_value' field to the end - however that won't be enough to shrink the struct size as Win32 ABI conventions mean that doubles must always be 8 bytes aligned; as as the struct contains a double it must be a multiple of 8 bytes in size. I'll adjust the test to test for <= 32 bytes in that case.

runrevmark commented:
Of course the most recent adjustment was also wrong - it should have been `#ifdef __32_BIT__` and not `#ifdef __32_BIT`. However padding to 32-bytes is still a good idea, it makes no difference to the amount of memory used (as all memory blocks will be a multiple of 8 or 16 bytes in size) and means that __MCString will always be 8 bytes aligned, even if they should ever be used as an array.

runrevmark committed:
[[ Bug 21993 ]] Tweak __MCName and __MCString layouts

This patch tweaks both the __MCName and __MCString struct layouts
to ensure they are more efficient.

The __MCName struct has had the now unused 'hash' member removed
when compiling for 64-bit architectures as this is stored in
free bits elsewhere in the structure.

The __MCString struct has had its layout changed when compiling
for 32-bit architectures so that the numeric_value double is
8-byte aligned, a padding member to ensure the struct is 32-bytes
in size on all platforms has also been added.

Additionally a low-level memory test has been added which verifies
the sizes of all the core __MCValue derived structs are the
expected (hand calculated and optimized for) values.

runrevmark commented:
@livecodeali : Hopefully this is correct now...

livecodeali commented:
@livecode-vulcan review ok fb7ab96

livecode-vulcan commented:
💙 review by @livecodeali ok fb7ab96dc1aba9944fe63e9d62db665405e264fc

livecode-vulcan commented:
😎 test success fb7ab96dc1aba9944fe63e9d62db665405e264fc

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/451)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/443)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/443)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/443)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/443)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1965)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/193)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/192)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1996)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1986)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1985)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2013)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1954)
